diff -ruN swarmops_org/Contexts/MethodContext.h SwarmOps/Contexts/MethodContext.h
--- swarmops_org/Contexts/MethodContext.h	2008-04-18 10:30:30.000000000 +0200
+++ SwarmOps/Contexts/MethodContext.h	2017-05-03 15:38:47.000000000 +0200
@@ -58,17 +58,17 @@
 	/* Make and return a context-struct.
 	 * Parameters correspond to the struct description above. */
 	struct SO_MethodContext SO_MakeMethodContext
-		(SO_FProblem f,
-		SO_FGradient fGradient,
-		void const* fContext,
-		SO_TDim fDim,
-		SO_TElm const* lowerInit,
-		SO_TElm const* upperInit,
-		SO_TElm const* lowerBound,
-		SO_TElm const* upperBound,
-		size_t numIterations,
-		void *settings,
-		struct SO_FitnessTrace *fitnessTrace);
+	(SO_FProblem f,
+	SO_FGradient fGradient,
+	void *fContext,
+	SO_TDim fDim,
+	SO_TElm const* lowerInit,
+	SO_TElm const* upperInit,
+	SO_TElm const* lowerBound,
+	SO_TElm const* upperBound,
+	size_t numIterations,
+	void *settings,
+	struct SO_FitnessTrace *fitnessTrace);
 
 	/* Free memory allocated by SO_MakeMethodContext(),
 	 * but not memory for the struct itself, the boundaries, etc. */
Binary files swarmops_org/libSwarmOps.so and SwarmOps/libSwarmOps.so differ
diff -ruN swarmops_org/Methods/DEEngine.c SwarmOps/Methods/DEEngine.c
--- swarmops_org/Methods/DEEngine.c	2008-07-28 11:25:38.000000000 +0200
+++ SwarmOps/Methods/DEEngine.c	2017-05-03 15:46:43.000000000 +0200
@@ -19,6 +19,14 @@
 
 /* ---------------------------------------------------------------- */
 
+const char*** SO_kNameDECrossover;
+
+/* ---------------------------------------------------------------- */
+
+
+
+/* ---------------------------------------------------------------- */
+
 /* Crossover for different DE variants. */
 void SO_DEEngineCrossover(	const size_t variant,			/* Which DE variant to use. */
 							SO_TDim n,						/* Dimensionality of problem. */
@@ -27,7 +35,7 @@
 							SO_TElm *x,						/* Destination vector. */
 							const SO_TElm *g,				/* Best vector. */
 							const size_t numAgents,			/* Number of agents. */
-							const SO_TElm **agents,			/* All agent vectors. */
+							SO_TElm **agents,			/* All agent vectors. */
 							struct RO_RandSet *randSet)		/* For picking random and distinct agents. */
 {
 	switch (variant)
@@ -56,7 +64,7 @@
 						SO_TElm *x,						/* Destination vector. */
 						const SO_TElm *g,				/* Best vector. */
 						const size_t numAgents,			/* Number of agents. */
-						const SO_TElm **agents,			/* All agent vectors. */
+						SO_TElm **agents,			/* All agent vectors. */
 						struct RO_RandSet *randSet)		/* For picking random and distinct agents. */
 {
 	/* Iteration variable. */
@@ -93,7 +101,7 @@
 						SO_TElm *x,						/* Destination vector. */
 						const SO_TElm *g,				/* Best vector. */
 						const size_t numAgents,			/* Number of agents. */
-						const SO_TElm **agents,			/* All agent vectors. */
+						SO_TElm **agents,			/* All agent vectors. */
 						struct RO_RandSet *randSet)		/* For picking random and distinct agents. */
 {
 	/* Iteration variable. */
@@ -133,7 +141,7 @@
 						const SO_TElm *F,				/* Differential weight. */
 						SO_TElm *x,						/* Destination vector. */
 						const size_t numAgents,			/* Number of agents. */
-						const SO_TElm **agents,			/* All agent vectors. */
+						SO_TElm **agents,			/* All agent vectors. */
 						struct RO_RandSet *randSet)		/* For picking random and distinct agents. */
 {
 	/* Iteration variable. */
@@ -170,7 +178,7 @@
 						const SO_TElm *F,				/* Differential weight. */
 						SO_TElm *x,						/* Destination vector. */
 						const size_t numAgents,			/* Number of agents. */
-						const SO_TElm **agents,			/* All agent vectors. */
+						SO_TElm **agents,			/* All agent vectors. */
 						struct RO_RandSet *randSet)		/* For picking random and distinct agents. */
 {
 	/* Iteration variable. */
diff -ruN swarmops_org/Methods/DEEngine.h SwarmOps/Methods/DEEngine.h
--- swarmops_org/Methods/DEEngine.h	2008-07-16 08:35:34.000000000 +0200
+++ SwarmOps/Methods/DEEngine.h	2017-05-03 15:38:46.000000000 +0200
@@ -39,12 +39,6 @@
 #define SO_kNameDECrossoverBest1Bin "best1bin"
 #define SO_kNameDECrossoverBest1BinSimple "best1binsimple"
 
-	/* ---------------------------------------------------------------- */
-
-	const char** SO_kNameDECrossover[];
-
-	/* ---------------------------------------------------------------- */
-
 	/* Crossover for different DE variants. */
 	void SO_DEEngineCrossover(	const size_t variant,			/* Which DE variant to use. */
 								SO_TDim n,						/* Dimensionality of problem. */
@@ -53,7 +47,7 @@
 								SO_TElm *x,						/* Destination vector. */
 								const SO_TElm *g,				/* Best vector. */
 								const size_t numAgents,			/* Number of agents. */
-								const SO_TElm **agents,			/* All agent vectors. */
+								SO_TElm **agents,			/* All agent vectors. */
 								struct RO_RandSet *randSet);	/* For picking random and distinct agents. */
 
 	/* ---------------------------------------------------------------- */
@@ -65,7 +59,7 @@
 								SO_TElm *x,						/* Destination vector. */
 								const SO_TElm *g,				/* Best vector. */
 								const size_t numAgents,			/* Number of agents. */
-								const SO_TElm **agents,			/* All agent vectors. */
+								SO_TElm **agents,			/* All agent vectors. */
 								struct RO_RandSet *randSet);	/* For picking random and distinct agents. */
 
 	/* ---------------------------------------------------------------- */
@@ -78,7 +72,7 @@
 								SO_TElm *x,						/* Destination vector. */
 								const SO_TElm *g,				/* Best vector. */
 								const size_t numAgents,			/* Number of agents. */
-								const SO_TElm **agents,			/* All agent vectors. */
+								SO_TElm **agents,			/* All agent vectors. */
 								struct RO_RandSet *randSet);	/* For picking random and distinct agents. */
 
 	/* ---------------------------------------------------------------- */
@@ -89,7 +83,7 @@
 								const SO_TElm *F,				/* Differential weight (vector). */
 								SO_TElm *x,						/* Destination vector. */
 								const size_t numAgents,			/* Number of agents. */
-								const SO_TElm **agents,			/* All agent vectors. */
+								SO_TElm **agents,			/* All agent vectors. */
 								struct RO_RandSet *randSet);	/* For picking random and distinct agents. */
 
 	/* ---------------------------------------------------------------- */
@@ -100,7 +94,7 @@
 								const SO_TElm *F,				/* Differential weight (vector). */
 								SO_TElm *x,						/* Destination vector. */
 								const size_t numAgents,			/* Number of agents. */
-								const SO_TElm **agents,			/* All agent vectors. */
+								SO_TElm **agents,			/* All agent vectors. */
 								struct RO_RandSet *randSet);	/* For picking random and distinct agents. */
 
 	/*----------------------------------------------------------------*/
diff -ruN swarmops_org/Methods/DE.h SwarmOps/Methods/DE.h
--- swarmops_org/Methods/DE.h	2008-07-28 14:05:24.000000000 +0200
+++ SwarmOps/Methods/DE.h	2017-05-03 15:38:46.000000000 +0200
@@ -56,7 +56,7 @@
 	/* ---------------------------------------------------------------- */
 
 	/* The optimization method. */
-	SO_TFitness SO_DE(const SO_TElm *param, void const* context, const SO_TFitness fitnessLimit);
+	SO_TFitness SO_DE(const SO_TElm *param, void * context, const SO_TFitness fitnessLimit);
 
 	/*----------------------------------------------------------------*/
 
diff -ruN swarmops_org/Methods/DESuite.h SwarmOps/Methods/DESuite.h
--- swarmops_org/Methods/DESuite.h	2008-07-28 13:48:24.000000000 +0200
+++ SwarmOps/Methods/DESuite.h	2017-05-03 15:38:46.000000000 +0200
@@ -85,7 +85,7 @@
 	/* ---------------------------------------------------------------- */
 
 	/* The optimization method. */
-	SO_TFitness SO_DESuite(const SO_TElm *param, void const* context, const SO_TFitness fitnessLimit);
+	SO_TFitness SO_DESuite(const SO_TElm *param, void * context, const SO_TFitness fitnessLimit);
 
 	/*----------------------------------------------------------------*/
 
diff -ruN swarmops_org/Methods/DETP.h SwarmOps/Methods/DETP.h
--- swarmops_org/Methods/DETP.h	2008-07-28 13:40:30.000000000 +0200
+++ SwarmOps/Methods/DETP.h	2017-05-03 15:38:46.000000000 +0200
@@ -73,7 +73,7 @@
 	/* ---------------------------------------------------------------- */
 
 	/* The optimization method. */
-	SO_TFitness SO_DETP(const SO_TElm *param, void const* context, const SO_TFitness fitnessLimit);
+	SO_TFitness SO_DETP(const SO_TElm *param, void * context, const SO_TFitness fitnessLimit);
 
 	/*----------------------------------------------------------------*/
 
diff -ruN swarmops_org/Methods/ELG.h SwarmOps/Methods/ELG.h
--- swarmops_org/Methods/ELG.h	2008-07-28 13:24:12.000000000 +0200
+++ SwarmOps/Methods/ELG.h	2017-05-03 15:38:46.000000000 +0200
@@ -48,7 +48,7 @@
 	/* ---------------------------------------------------------------- */
 
 	/* The optimization method. */
-	SO_TFitness SO_ELG(const SO_TElm *param, void const* context, const SO_TFitness fitnessLimit);
+	SO_TFitness SO_ELG(const SO_TElm *param, void * context, const SO_TFitness fitnessLimit);
 
 	/*----------------------------------------------------------------*/
 
diff -ruN swarmops_org/Methods/FAE.h SwarmOps/Methods/FAE.h
--- swarmops_org/Methods/FAE.h	2008-07-28 13:24:06.000000000 +0200
+++ SwarmOps/Methods/FAE.h	2017-05-03 15:38:46.000000000 +0200
@@ -50,7 +50,7 @@
 	/* ---------------------------------------------------------------- */
 
 	/* The optimization method. */
-	SO_TFitness SO_FAE(const SO_TElm *param, void const* context, const SO_TFitness fitnessLimit);
+	SO_TFitness SO_FAE(const SO_TElm *param, void * context, const SO_TFitness fitnessLimit);
 
 	/*----------------------------------------------------------------*/
 
Binary files swarmops_org/Methods/Helpers/LooperLog.o and SwarmOps/Methods/Helpers/LooperLog.o differ
Binary files swarmops_org/Methods/Helpers/Looper.o and SwarmOps/Methods/Helpers/Looper.o differ
Binary files swarmops_org/Methods/Helpers/Multi.o and SwarmOps/Methods/Helpers/Multi.o differ
Binary files swarmops_org/Methods/Helpers/Printer.o and SwarmOps/Methods/Helpers/Printer.o differ
diff -ruN swarmops_org/Methods/JDE.h SwarmOps/Methods/JDE.h
--- swarmops_org/Methods/JDE.h	2008-07-28 13:23:28.000000000 +0200
+++ SwarmOps/Methods/JDE.h	2017-05-03 15:38:46.000000000 +0200
@@ -87,7 +87,7 @@
 	/* ---------------------------------------------------------------- */
 
 	/* The optimization method. */
-	SO_TFitness SO_JDE(const SO_TElm *param, void const* context, const SO_TFitness fitnessLimit);
+	SO_TFitness SO_JDE(const SO_TElm *param, void * context, const SO_TFitness fitnessLimit);
 
 	/*----------------------------------------------------------------*/
 
diff -ruN swarmops_org/Methods/MOL.h SwarmOps/Methods/MOL.h
--- swarmops_org/Methods/MOL.h	2008-07-28 13:23:08.000000000 +0200
+++ SwarmOps/Methods/MOL.h	2017-05-03 15:38:47.000000000 +0200
@@ -60,7 +60,7 @@
 	/* ---------------------------------------------------------------- */
 
 	/* The optimization method. */
-	SO_TFitness SO_MOL(const SO_TElm *param, void const* context, const SO_TFitness fitnessLimit);
+	SO_TFitness SO_MOL(const SO_TElm *param, void * context, const SO_TFitness fitnessLimit);
 
 	/*----------------------------------------------------------------*/
 
diff -ruN swarmops_org/Methods/MYG.h SwarmOps/Methods/MYG.h
--- swarmops_org/Methods/MYG.h	2008-07-28 14:42:26.000000000 +0200
+++ SwarmOps/Methods/MYG.h	2017-05-03 15:38:47.000000000 +0200
@@ -56,7 +56,7 @@
 	/* ---------------------------------------------------------------- */
 
 	/* The optimization method. */
-	SO_TFitness SO_MYG(const SO_TElm *param, void const* context, const SO_TFitness fitnessLimit);
+	SO_TFitness SO_MYG(const SO_TElm *param, void * context, const SO_TFitness fitnessLimit);
 
 	/*----------------------------------------------------------------*/
 
diff -ruN swarmops_org/Methods/PSO.h SwarmOps/Methods/PSO.h
--- swarmops_org/Methods/PSO.h	2008-07-28 13:22:48.000000000 +0200
+++ SwarmOps/Methods/PSO.h	2017-05-03 15:38:47.000000000 +0200
@@ -59,7 +59,7 @@
 	/* ---------------------------------------------------------------- */
 
 	/* The optimization method. */
-	SO_TFitness SO_PSO(const SO_TElm *param, void const* context, const SO_TFitness fitnessLimit);
+	SO_TFitness SO_PSO(const SO_TElm *param, void * context, const SO_TFitness fitnessLimit);
 
 	/*----------------------------------------------------------------*/
 
diff -ruN swarmops_org/RandomOps/Makefile SwarmOps/RandomOps/Makefile
--- swarmops_org/RandomOps/Makefile	1970-01-01 01:00:00.000000000 +0100
+++ SwarmOps/RandomOps/Makefile	2014-12-11 15:25:15.000000000 +0100
@@ -0,0 +1,19 @@
+CC=g++
+CPPFLAGS :=             -Wall -I. -I./../
+
+# determine all sources and from that all targets
+sources :=              $(wildcard *.c) $(wildcard */*.c)
+objects=$(sources:.c=.o)
+
+# build all and strip programs afterwards 
+all:    $(objects) librandom.a
+
+librandom.a: 
+	ar -cvq $@ $(objects)
+
+.c.o: 
+	$(CC) $(CPPFLAGS) -c $^ -o $@
+
+
+clean:
+	rm -f $(objects)
diff -ruN swarmops_org/Statistics/FitnessTrace.c SwarmOps/Statistics/FitnessTrace.c
--- swarmops_org/Statistics/FitnessTrace.c	2008-04-03 09:35:40.000000000 +0200
+++ SwarmOps/Statistics/FitnessTrace.c	2017-05-03 15:38:46.000000000 +0200
@@ -84,13 +84,12 @@
 	if (SO_UseFitnessTrace(trc))
 	{
 		size_t i;
-		errno_t err;
 		FILE *stream;
 
 		/* Create file. */
-		err = fopen_s(&stream, trc->filename, "w");
+		stream = fopen(trc->filename, "w");
 
-		if (err == 0)
+		if (stream != 0)
 		{
 			/* Write trace to file. */
 			for (i=0; i<trc->numIterations; i++)
diff -ruN swarmops_org/Tools/Bound.c SwarmOps/Tools/Bound.c
--- swarmops_org/Tools/Bound.c	2008-04-03 09:53:16.000000000 +0200
+++ SwarmOps/Tools/Bound.c	2017-05-03 15:51:20.000000000 +0200
@@ -12,6 +12,7 @@
  *
  * ================================================================ */
 
+#include <SwarmOps/Tools/Bound.h>
 #include <SwarmOps/Tools/Types.h>
 #include <assert.h>
 
diff -ruN swarmops_org/Tools/Denormal.c SwarmOps/Tools/Denormal.c
--- swarmops_org/Tools/Denormal.c	2008-04-02 09:09:40.000000000 +0200
+++ SwarmOps/Tools/Denormal.c	2017-05-03 15:38:46.000000000 +0200
@@ -12,6 +12,7 @@
  *
  * ================================================================ */
 
+#include <SwarmOps/Tools/Denormal.h>
 #include <SwarmOps/Tools/Types.h>
 #include <math.h>
 #include <assert.h>
diff -ruN swarmops_org/Tools/Error.c SwarmOps/Tools/Error.c
--- swarmops_org/Tools/Error.c	2008-04-03 07:45:58.000000000 +0200
+++ SwarmOps/Tools/Error.c	2014-12-11 15:25:15.000000000 +0100
@@ -12,6 +12,7 @@
  *
  * ================================================================ */
 
+#include "Tools/Error.h"
 #include <stdio.h>
 #include <stdlib.h>
 #include <assert.h>
@@ -19,8 +20,7 @@
 /* ---------------------------------------------------------------- */
 
 /* The error messages. */
-
-const char SO_kErrMsgMemory[] = "Out of memory.";
+const char * SO_kErrMsgMemory = "Out of memory.";
 
 /* ---------------------------------------------------------------- */
 
diff -ruN swarmops_org/Tools/Error.h SwarmOps/Tools/Error.h
--- swarmops_org/Tools/Error.h	2008-04-03 07:47:10.000000000 +0200
+++ SwarmOps/Tools/Error.h	2017-05-03 15:53:09.000000000 +0200
@@ -34,7 +34,7 @@
 
 	/*----------------------------------------------------------------*/
 
-	const char SO_kErrMsgMemory[];		/* Out of memory message. */
+	extern const char * SO_kErrMsgMemory;		/* Out of memory message. */
 
 	/*----------------------------------------------------------------*/
 
diff -ruN swarmops_org/Tools/Memory.c SwarmOps/Tools/Memory.c
--- swarmops_org/Tools/Memory.c	2008-04-03 08:07:50.000000000 +0200
+++ SwarmOps/Tools/Memory.c	2017-05-03 15:38:46.000000000 +0200
@@ -13,7 +13,6 @@
  * ================================================================ */
 
 #include <SwarmOps/Tools/Memory.h>
-#include <SwarmOps/Tools/Error.h>
 #include <stdlib.h>
 #include <assert.h>
 
diff -ruN swarmops_org/Tools/Memory.h SwarmOps/Tools/Memory.h
--- swarmops_org/Tools/Memory.h	2008-04-03 08:03:44.000000000 +0200
+++ SwarmOps/Tools/Memory.h	2017-05-03 15:38:46.000000000 +0200
@@ -15,6 +15,7 @@
 #ifndef SO_MEMORY_H
 #define SO_MEMORY_H
 
+#include <SwarmOps/Tools/Error.h>
 #include <stddef.h>
 
 #ifdef  __cplusplus
diff -ruN swarmops_org/Tools/Random.c SwarmOps/Tools/Random.c
--- swarmops_org/Tools/Random.c	2008-04-03 08:11:38.000000000 +0200
+++ SwarmOps/Tools/Random.c	2017-05-03 15:46:43.000000000 +0200
@@ -12,7 +12,8 @@
  *
  * ================================================================ */
 
-#include <RandomOps/random.h>
+#include <SwarmOps/Tools/Random.h>
+#include <RandomOps/Random.h>
 
 /* ---------------------------------------------------------------- */
 
diff -ruN swarmops_org/Tools/Random.h SwarmOps/Tools/Random.h
--- swarmops_org/Tools/Random.h	2008-04-03 08:11:02.000000000 +0200
+++ SwarmOps/Tools/Random.h	2017-05-03 15:38:47.000000000 +0200
@@ -17,6 +17,8 @@
 #ifndef SO_RANDOM_H
 #define SO_RANDOM_H
 
+#include <SwarmOps/Tools/Types.h>
+
 #ifdef  __cplusplus
 extern "C" {
 #endif
